File: 962. Maximum Width Ramp/max_width_ramp.rs
use std::cmp;


impl Solution {
    
    fn find_furthest_min(target: i32, nums: &Vec<i32>, stack: &Vec<usize>) -> usize {
        let mut lo = 0 as usize;
        let mut hi = stack.len()-1;
        let mut ans = 0 as usize;
        
        while lo <= hi && hi < stack.len() {
            let mid = lo + (hi-lo)/2;
            if nums[stack[mid]] <= target {
                ans = stack[mid];
                hi = mid-1;
            } else {
                lo = mid+1;
            }
        }
        return ans
    }
    
    
    // O(nlogn) time,
    // O(n) space,
    // Approach: monotonic stack, binary search, 
    pub fn max_width_ramp(nums: Vec<i32>) -> i32 {
        let mut dec_stack = Vec::new();
        let mut max_width = 0;
        
        for i in 0..nums.len() {
            let stack_len = dec_stack.len();
            if stack_len == 0 || nums[dec_stack[stack_len-1]] > nums[i] {
                dec_stack.push(i);
            } else {
                let furthest_min_index = Solution::find_furthest_min(nums[i], &nums, &dec_stack);
                max_width = cmp::max(max_width, i-furthest_min_index);
            }
        }
        return max_width as i32;
    }
}

File: 962. Maximum Width Ramp/maxWidthRamp.go
package MaxWidthRamp

import "math"

// O(n) time, 2n practically
// O(n) space,
// Approach: stack, monotonic stack
func maxWidthRamp(nums []int) int {
    
    stack := []int{}
    
    for i, val := range nums {
        if len(stack) == 0 || nums[stack[len(stack)-1]] >= val {
            stack = append(stack, i)
        }
    }
    
    maxWidth := 0
    for i := len(nums)-1; i >= 0; i-- {
        for len(stack) > 0 && nums[i] >= nums[stack[len(stack)-1]] {
            rampStartIndex := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            currWidth := float64(i-rampStartIndex)
            maxWidth = int(math.Max(float64(maxWidth), currWidth))
        }
    }
    
    return maxWidth
}

// O(n) time, 2n practically
// O(n) space,
// Approach: two pointers
func maxWidthRamp2(nums []int) int {
    
    // [8,8,8,5,5,5]
    rightMax := make([]int, len(nums))
    rightMax[len(nums)-1] = nums[len(nums)-1]
    for i := len(nums)-2; i >= 0; i-- {
        if rightMax[i+1] > nums[i] {
            rightMax[i] = rightMax[i+1]
        } else {
            rightMax[i] = nums[i]
        }
    }
    
    maxWidth := float64(0)
    left, right := 0, 0
    for right < len(nums) {
        for nums[left] > rightMax[right] {
            left += 1
        }
        currWidth := float64(right-left)
        maxWidth = math.Max(maxWidth, currWidth)
        right += 1
    }
    
    return int(maxWidth)
}
